cds/utils.bal

```diff
@@ -0,0 +1,192 @@
+    map<string> prefetchTemplates = {};
+    map<r4:DomainResource> prefetchData = {};
+    CdsRequest request = cdsRequest.clone();
+
+    // If the CDS service has no prefetch templates defined, then no need to validate the request
+    if cdsService.prefetch is map<string> {
+        prefetchTemplates = <map<string>>cdsService.prefetch;
+        if (prefetchTemplates.keys().length() < 1) {
+            return request;
+        }
+    } else {
+        return request;
+    }
+
+    if (request.prefetch is map<r4:DomainResource>) {
+        prefetchData = <map<r4:DomainResource>>request.prefetch;
+    }
+
+    // If there are no any prefetch data in the request, or
+    // required prefetch template keys are not matching with keys in the request prefetch data 
+    //then following should we given in the request to fetch the required FHIR data
+    string[] templateKeys = prefetchTemplates.keys().sort();
+    string[] dataKeys = prefetchData.keys().sort();
+    if (templateKeys != dataKeys || templateKeys.count() > dataKeys.count()) {
+        // 1. Fhir server url
+        if (request.fhirServer !is string) {
+            return createCdsError("Can not find fhirServer url in the request", 400);
+        }
+
+        // 2. FhirAuthorization object contains access token, scopes etc
+        if (request.fhirAuthorization !is FhirAuthorization) {
+            return createCdsError("Can not find fhirAuthorization in the request", 400);
+        }
+
+        // Fetch missing prefetch FHIR data
+        foreach string templateId in prefetchTemplates.keys() {
+            if (!prefetchData.hasKey(templateId)) {
+                request = check fetchFhirResource(templateId, prefetchTemplates.get(templateId), request);
+            } else {
+                r4:FHIRValidationError? validated = validator:validate(prefetchData.get(templateId));
+                if validated is r4:FHIRValidationError {
+                    return createCdsError(string `FHIR data provided for prefetch ${templateId} in the request is not valid`, 412, cause = validated);
+                }
+            }
+        }
+    } else {
+        foreach var key in prefetchData.keys() {
+            r4:FHIRValidationError? validated = validator:validate(prefetchData.get(key));
+            if validated is r4:FHIRValidationError {
+                return createCdsError(string `FHIR data provided for the prefetch: ${key}`, 400, cause = validated);
+            }
+        }
+    }
+
+    return request;
+}
+
+isolated function fetchFhirResource(string prefetchTemplateKey, string prefetchTemplate, CdsRequest cdsRequest) returns CdsError|CdsRequest {
+    string template = prefetchTemplate.clone();
+    string split = regexp:split(re `\.`, regexp:split(re `\}\}`, regexp:split(re `\{\{`, prefetchTemplate)[1])[0])[1];
+    string fhirServer = <string>cdsRequest.fhirServer;
+    anydata contextValue = cdsRequest.context.get(split);
+    if contextValue is string {
```

Comment:
Can't we process these at the init withour processing for each request?

Reply:
Improved